### PR TITLE
[dbsp] schedule circuits using tokio.

### LINKED ISSUE #21: Replace `UnsafeCell` with `RefCell` in `circuit_builder.rs`
<a href="https://github.com/ryzhyk"></img></a> **Issue by [ryzhyk](https://github.com/ryzhyk)**
_Fri Jun  3 06:53:19 2022_
_Originally opened as https://github.com/vmware/database-stream-processor/issues/76_

----

### Is your feature request related to a problem? Please describe.

`UnsafeCell` was a premature optimization. Its use is not well encapsulated, and it likely enables unsafe behavior.

### Describe the solution you'd like

-

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### LINKED ISSUE #3287: [dbsp] support for async IO and cooperative scheduling
We are adding support for async IO to DBSP, where an operator can start a long-running I/O operation (e.g., prefetching a batch of keys) and get notified about completion via callback. To take full advantage of this we need to be able to schedule other runnable operators from the same circuit on the same worker thread in the meantime. In order to enable this without significantly complicating the implementation of operators, I suggest we make the operator's `eval()` method `async` and use the single-threaded `tokio` runtime to cooperatively schedule operators on the DBSP worker thread.

### PR DESCRIPTION
**Pull request by ryzhyk**
The DBSP scheduler used to run operators one by one, executing each operator to completion before picking the next runnable operator.  This is inefficient for I/O intensive operators, as it doesn't allow parallelizing I/O and compute or I/O from multiple operators (within one worker thread).  As we move toward using high-latency storage

This commit enables cooperative scheduling of operators by running them within a single-threaded tokio runtime. Specifically, it makes the following changes:

* Operator `eval` methods are now async, enabling an operator to yield CPU to other runnable operators.

* The scheduler no longer chooses the next operator to run; instead it simply spawns a task for each operator once all of its predecessors have been evaluated.

* We have to be more careful with managing shared references to nodes and edges of the circuit graph in circuit_builder.rs, as there can be multiple simultaneously running operators that read from common streams.  To make it easier to catch bugs, I finally switched from using UnsafeCell to RefCell (issue https://github.com/feldera/feldera/issues/21).

None of the existing operators currently take advantage of async scheduling; however this commit enables the creation of such operators, which are coming to DBSP shortly.

Based on nexmark Rust benchmarks, this change doesn't have significant impact on performance.

Resolves #21 
Resolves #3287

### CODE REVIEW

File: crates/dbsp/src/operator/dynamic/controlled_filter.rs [hunk 1]
```diff
@@ -271,7 +271,7 @@ where
             cursor.step_key();
         }
 
-        unsafe { self.error_stream_val.put(errors.done()) };
+        self.error_stream_val.put(errors.done());
```
Comment (mihaibudiu) [inline crates/dbsp/src/operator/dynamic/controlled_filter.rs:274]: this change is surprising

File: crates/dbsp/src/operator/condition.rs [hunk 2]
```diff
@@ -141,18 +141,18 @@ impl<C> Condition<C> {
...
         typed_batch::{OrdIndexedZSet, OrdZSet},
         utils::Tup2,
         zset, ChildCircuit, Circuit, RootCircuit, Stream,
     };
 
-    #[test]
-    fn iterate_with_conditions_static() {
-        iterate_with_conditions::<StaticScheduler>();
-    }
+    // #[test]
```
Comment (mihaibudiu) [inline crates/dbsp/src/operator/condition.rs:152]: delete?

File: crates/dbsp/src/operator/communication/exchange.rs [hunk 3]
```diff
@@ -1092,11 +1092,11 @@ mod tests {
         hruntime.join().unwrap();
     }
 
-    #[test]
-    #[cfg_attr(miri, ignore)]
-    fn test_exchange_operators_static() {
-        test_exchange_operators::<StaticScheduler>();
-    }
+    // #[test]
```
Comment (mihaibudiu) [inline crates/dbsp/src/operator/communication/exchange.rs:1095]: delete?

File: crates/dbsp/src/circuit/schedule/mod.rs [hunk 4]
```diff
@@ -8,11 +8,13 @@ use std::{
...
     error::Error as StdError,
     fmt::{Display, Error as FmtError, Formatter},
+    future::Future,
+    pin::Pin,
     string::ToString,
 };
 
-mod static_scheduler;
-pub use static_scheduler::StaticScheduler;
+// mod static_scheduler;
```
Comment (mihaibudiu) [inline crates/dbsp/src/circuit/schedule/mod.rs:16]: delete?

File: crates/dbsp/src/circuit/schedule/mod.rs [hunk 5]
```diff
@@ -143,20 +152,23 @@ where
...
-        circuit.log_scheduler_event(&SchedulerEvent::clock_start());
-        circuit.clock_start(0);
-
-        loop {
-            self.scheduler.step(circuit)?;
-            if (self.termination_check)()? {
-                break;
+    fn run<'a>(&'a self, circuit: &C) -> Pin<Box<dyn Future<Output = Result<(), Error>> + 'a>> {
+        let circuit = circuit.clone();
+        Box::pin(async move {
```
Comment (mihaibudiu) [inline crates/dbsp/src/circuit/schedule/mod.rs:157]: never saw this before.

File: crates/dbsp/src/circuit/schedule/dynamic_scheduler.rs [hunk 6]
```diff
@@ -319,62 +286,120 @@ impl Inner {
...
-                return Err(Error::Killed);
-            }
-
-            match self.dequeue_next_task() {
-                None => {
-                    // No more tasks in the run queue -- try to add some by
-                    // processing notifications.
-                    self.process_notifications(circuit);
+        loop {
+            select! {
```
Comment (mihaibudiu) [inline crates/dbsp/src/circuit/schedule/dynamic_scheduler.rs:347]: this macro looks like magic, can't figure out how it knows what to wait on

File: crates/dbsp/src/circuit/circuit_builder.rs [hunk 7]
```diff
@@ -1355,7 +1389,7 @@ pub trait Circuit: WithClock + Clone + 'static {
     /// Evaluate operator with the given id.
     ///
     /// This method should only be used by schedulers.
-    fn eval_node(&self, id: NodeId) -> Result<(), SchedulerError>;
+    fn eval_node(&self, id: NodeId) -> impl Future<Output = Result<(), SchedulerError>>;
```
Comment (mihaibudiu) [inline crates/dbsp/src/circuit/circuit_builder.rs:1392]: will this change when we add error handling?

File: crates/dbsp/src/circuit/circuit_builder.rs [hunk 8]
```diff
@@ -2045,14 +2081,14 @@ impl<P> ChildCircuit<P>
 where
     P: WithClock,
 {
-    /// Mutably borrow the inner circuit.
-    fn inner_mut(&self) -> RefMut<'_, CircuitInner<P>> {
-        self.inner.borrow_mut()
-    }
+    // /// Mutably borrow the inner circuit.
```
Comment (mihaibudiu) [inline crates/dbsp/src/circuit/circuit_builder.rs:2084]: delete?

File: crates/dbsp/src/circuit/circuit_builder.rs [hunk 9]
```diff
@@ -2425,15 +2477,19 @@ where
...
+                store.get_mut(&key).unwrap()
+            });
         }
 
         let new = f();
 
         // TODO: Use `RefMut::filter_map()` to only perform one lookup in the happy path
         //       https://github.com/rust-lang/rust/issues/81061
-        RefMut::map(self.inner_mut(), |c| c.store.entry(key).or_insert(new))
+        RefMut::map(self.inner().store.borrow_mut(), |store| {
```
Comment (mihaibudiu) [inline crates/dbsp/src/circuit/circuit_builder.rs:2490]: can the comment above be resolved too?

File: crates/dbsp/src/circuit/circuit_builder.rs [hunk 10]
```diff
@@ -3724,24 +3807,71 @@ where
...
+                            Cow::Borrowed(StreamValue::peek(&val1)),
+                            Cow::Borrowed(StreamValue::peek(&val2)),
+                        )
+                        .await;
+                }
 
-        let input2 = self.input_stream2.take();
+                self.input_stream1.get_mut().consume_token();
+                self.input_stream2.get_mut().consume_token();
+            } else {
```
Comment (mihaibudiu) [inline crates/dbsp/src/circuit/circuit_builder.rs:3828]: this looks complicated, isn't is simpler just to process each input separately?

File: crates/dbsp/src/circuit/circuit_builder.rs [hunk 11]
```diff
@@ -3930,17 +4083,14 @@ where
...
-        let is_alias2 = input_stream2.ptr_eq(&input_stream3);
         Self {
             id: circuit.global_node_id().child(id),
             operator,
             input_stream1,
             input_stream2,
             input_stream3,
-            is_alias1,
-            is_alias2,
+            // is_alias1,
```
Comment (mihaibudiu) [inline crates/dbsp/src/circuit/circuit_builder.rs:4092]: delete?

File: crates/dbsp/src/circuit/circuit_builder.rs [hunk 12]
```diff
@@ -3983,31 +4133,35 @@ where
...
-            self.input_stream2.take()
-        };
+    // Justification: see StreamValue::take() comment.
+    #[allow(clippy::await_holding_refcell_ref)]
+    fn eval<'a>(&'a mut self) -> Pin<Box<dyn Future<Output = Result<(), SchedulerError>> + 'a>> {
+        Box::pin(async {
+            {
+                self.output_stream.put(
+                    self.operator
+                        .eval(
```
Comment (mihaibudiu) [inline crates/dbsp/src/circuit/circuit_builder.rs:4143]: why is this one so much simpler?

File: crates/dbsp/src/circuit/circuit_builder.rs [hunk 13]
```diff
@@ -4044,13 +4198,13 @@ struct QuaternaryNode<C, I1, I2, I3, I4, O, Op> {
...
     input_stream4: Stream<C, I4>,
     output_stream: Stream<C, O>,
-    // `true` if `input_stream1` is an alias to `input_stream2`, `input_stream3` or
-    // `input_stream4`.
-    is_alias1: bool,
-    // `true` if `input_stream2` is an alias to `input_stream3` or `input_stream4`.
-    is_alias2: bool,
-    // `true` if `input_stream3` is an alias to `input_stream4`.
-    is_alias3: bool,
+    // // `true` if `input_stream1` is an alias to `input_stream2`, `input_stream3` or
```
Comment (mihaibudiu) [inline crates/dbsp/src/circuit/circuit_builder.rs:4201]: delete?

File: crates/dbsp/src/circuit/circuit_builder.rs [hunk 14]
```diff
@@ -4071,22 +4225,22 @@ where
         circuit: C,
         id: NodeId,
     ) -> Self {
-        let is_alias1 = input_stream1.ptr_eq(&input_stream2)
-            || input_stream1.ptr_eq(&input_stream3)
-            || input_stream1.ptr_eq(&input_stream4);
-        let is_alias2 =
-            input_stream2.ptr_eq(&input_stream3) || input_stream2.ptr_eq(&input_stream4);
-        let is_alias3 = input_stream3.ptr_eq(&input_stream4);
+        // let is_alias1 = input_stream1.ptr_eq(&input_stream2)
```
Comment (mihaibudiu) [inline crates/dbsp/src/circuit/circuit_builder.rs:4228]: delete?

File: crates/dbsp/src/circuit/circuit_builder.rs [hunk 15]
```diff
@@ -4071,22 +4225,22 @@ where
...
             id: circuit.global_node_id().child(id),
             operator,
             input_stream1,
             input_stream2,
             input_stream3,
             input_stream4,
-            is_alias1,
-            is_alias2,
-            is_alias3,
+            // is_alias1,
```
Comment (mihaibudiu) [inline crates/dbsp/src/circuit/circuit_builder.rs:4241]: delete?

File: crates/dbsp/src/circuit/circuit_builder.rs [hunk 16]
```diff
@@ -4213,27 +4367,24 @@ where
...
-                if input_streams[i].0.ptr_eq(&input_streams[j].0) {
-                    input_streams[i].1 = true;
-                    aliases.push(i);
-                    break;
-                }
-            }
-        }
-        aliases.shrink_to_fit();
+        let mut input_streams: Vec<_> = input_streams.into_iter().collect();
+        // let mut aliases = Vec::new();
```
Comment (mihaibudiu) [inline crates/dbsp/src/circuit/circuit_builder.rs:4371]: delete?

File: crates/dbsp/src/circuit/circuit_builder.rs [hunk 17]
```diff
@@ -4197,9 +4351,9 @@ where
     operator: Op,
     // The second field of the tuple indicates if the stream is an
     // alias to an earlier stream.
-    input_streams: Vec<(Stream<C, I>, bool)>,
-    // Streams that are aliases.
-    aliases: Vec<usize>,
+    input_streams: Vec<Stream<C, I>>,
+    // // Streams that are aliases.
```
Comment (mihaibudiu) [inline crates/dbsp/src/circuit/circuit_builder.rs:4355]: delete?

File: crates/dbsp/src/circuit/circuit_builder.rs [hunk 18]
```diff
@@ -4213,27 +4367,24 @@ where
...
+        //     }
+        // }
+        //aliases.shrink_to_fit();
         input_streams.shrink_to_fit();
         Self {
             id: circuit.global_node_id().child(id),
             operator,
             input_streams,
-            aliases,
+            //aliases,
```
Comment (mihaibudiu) [inline crates/dbsp/src/circuit/circuit_builder.rs:4387]: delete?

File: crates/dbsp/src/circuit/circuit_builder.rs [hunk 19]
```diff
@@ -4274,31 +4425,35 @@ where
         self.operator.register_ready_callback(cb);
     }
 
-    unsafe fn eval(&mut self) -> Result<(), SchedulerError> {
-        self.output_stream
-            .put(
+    fn eval<'a>(&'a mut self) -> Pin<Box<dyn Future<Output = Result<(), SchedulerError>> + 'a>> {
+        Box::pin(async {
+            // TODO: preallocate
```
Comment (mihaibudiu) [inline crates/dbsp/src/circuit/circuit_builder.rs:4430]: why TODO and not just do it now?

File: crates/dbsp/src/circuit/circuit_builder.rs [hunk 20]
```diff
@@ -4739,7 +4913,12 @@ impl CircuitHandle {
     pub fn step(&self) -> Result<(), SchedulerError> {
         // TODO: Add a runtime check to prevent re-entering this method from an
         // operator.
-        self.executor.run(&self.circuit)
+        self.tokio_runtime.block_on(async {
```
Comment (mihaibudiu) [inline crates/dbsp/src/circuit/circuit_builder.rs:4916]: can you solve the comment above too?

File: crates/dbsp/src/circuit/circuit_builder.rs [hunk 21]
```diff
@@ -4805,19 +4984,19 @@ impl CircuitHandle {
...
     };
     use anyhow::anyhow;
     use std::{cell::RefCell, ops::Deref, rc::Rc, vec::Vec};
 
-    // Compute the sum of numbers from 0 to 99.
-    #[test]
-    fn sum_circuit_static() {
-        sum_circuit::<StaticScheduler>();
-    }
+    // // Compute the sum of numbers from 0 to 99.
```
Comment (mihaibudiu) [inline crates/dbsp/src/circuit/circuit_builder.rs:4995]: delete all the static scheduler tests?

**Review by mihaibudiu — APPROVED** [review]
this looks like a fairly mechanical change, which deletes a lot of code and removes some unsafe. What is there not to like?

File: crates/dbsp/src/circuit/circuit_builder.rs [hunk 22]
```diff
@@ -745,41 +796,24 @@ impl<C, D> Stream<C, D>
...
-        val.val.as_ref().unwrap()
+    fn val(&self) -> &RefCell<StreamValue<D>> {
+        &self.val.0
     }
 
     /// Puts a value in the stream, overwriting the previous value if any.
     ///
     /// #Safety
     ///
     /// The caller must have exclusive access to the current stream.
```
Comment (blp) [inline crates/dbsp/src/circuit/circuit_builder.rs:815]: I guess the safety comment can be deleted since this is no longer unsafe.

File: crates/dbsp/src/circuit/circuit_builder.rs [hunk 23]
```diff
@@ -130,15 +181,15 @@ impl<D> RefStreamValue<D> {
     ///
     /// Noone should be holding a reference to the stream value, e.g., the stream cannot
     /// simultaneously be an input and an output of the same operator.
```
Comment (blp) [inline crates/dbsp/src/circuit/circuit_builder.rs:183]: I think this safety comment can be deleted too.

File: crates/dbsp/src/circuit/schedule/dynamic_scheduler.rs [hunk 24]
```diff
@@ -319,62 +286,120 @@ impl Inner {
...
 
-    fn step<C>(&mut self, circuit: &C) -> Result<(), Error>
+    /// Spawn a tokio task to evaluate `node_id`.
+    fn spawn_task<C>(&mut self, circuit: &C, node_id: NodeId)
     where
         C: Circuit,
     {
-        circuit.log_scheduler_event(&SchedulerEvent::step_start(circuit.global_id().deref()));
+        let task = &mut self.tasks[node_id.id()];
+        debug_assert!(task.unsatisfied_dependencies == 0);
```
Comment (blp) [inline crates/dbsp/src/circuit/schedule/dynamic_scheduler.rs:295]:
```suggestion
        debug_assert_eq!(task.unsatisfied_dependencies, 0);
```

File: crates/dbsp/src/circuit/schedule/dynamic_scheduler.rs [hunk 25]
```diff
@@ -319,62 +286,120 @@ impl Inner {
...
+        let circuit = circuit.clone();
+
+        self.handles.spawn_local(async move {
+            let result = circuit.eval_node(node_id).await;
+            (node_id, result)
+        });
+    }
+
+    async fn abort(&mut self) {
+        // wait for all started tasks to abort
```
Comment (blp) [inline crates/dbsp/src/circuit/schedule/dynamic_scheduler.rs:315]: We could cancel the tasks instead of just waiting for them to finish, by adding:

```
        self.handles.abort_all();
```

**Review by blp — APPROVED** [review]
I'm much happier with RefCell than UnsafeCell.

This change removes a delightful amount of unsafe.

Like Mihai, I support deleting  the commented-out code.

File: crates/dbsp/src/operator/dynamic/controlled_filter.rs (hunk 1 above)
Comment (ryzhyk) [reply crates/dbsp/src/operator/dynamic/controlled_filter.rs:274]: it used to be unsafe, but now is safe :)

File: crates/dbsp/src/circuit/schedule/dynamic_scheduler.rs (hunk 25 above)
Comment (blp) [reply crates/dbsp/src/circuit/schedule/dynamic_scheduler.rs]: I wasn't sure what was the best choice here. It's OK with me not to cancel them.
